google/datalab/contrib/bigquery/commands/_bigquery.py

```diff
@@ -0,0 +1,295 @@
+    """Implements the pipeline subcommand in the %%bq magic.
+
+    The supported syntax is:
+
+        %%bq pipeline <args>
+        [<inline YAML>]
+
+    Args:
+      args: the arguments following '%%bq pipeline'.
+      cell_body: the contents of the cell
+    """
+    name = args.get('name')
+    if name is None:
+        raise Exception("Pipeline name was not specified.")
+
+    bq_pipeline_config = google.datalab.utils.commands.parse_config(
+        cell_body, google.datalab.utils.commands.notebook_environment())
+
+    load_task_config_name = 'bq_pipeline_load_task'
+    load_task_config = {'type': 'pydatalab.bq.load'}
+    _add_load_parameters(load_task_config, bq_pipeline_config.get('input', None))
+
+    execute_task_config_name = 'bq_pipeline_execute_task'
+    execute_task_config = {'type': 'pydatalab.bq.execute', 'up_stream': [load_task_config_name]}
+    _add_execute_parameters(execute_task_config, bq_pipeline_config['transformation'])
+
+    extract_task_config_name = 'bq_pipeline_extract_task'
+    extract_task_config = {'type': 'pydatalab.bq.extract', 'up_stream': [execute_task_config_name]}
+    _add_extract_parameters(extract_task_config, execute_task_config, bq_pipeline_config['output'])
+
+    pipeline_spec = {
+        'email': bq_pipeline_config['email'],
+        'schedule': bq_pipeline_config['schedule'],
+    }
+
+    # These sections are only set when they aren't None
+    pipeline_spec['tasks'] = {}
+    if load_task_config:
+        pipeline_spec['tasks'][load_task_config_name] = load_task_config
+    if execute_task_config:
+        pipeline_spec['tasks'][execute_task_config_name] = execute_task_config
+    if extract_task_config:
+        pipeline_spec['tasks'][extract_task_config_name] = extract_task_config
+
+    if not load_task_config and not execute_task_config and not extract_task_config:
+        raise Exception('Pipeline has no tasks to execute.')
+
+    pipeline = google.datalab.contrib.pipeline._pipeline.Pipeline(name, pipeline_spec)
+    google.datalab.utils.commands.notebook_environment()[name] = pipeline
+
+    debug = args.get('debug')
+    if debug is True:
+        return pipeline.py
+
+
+def _add_load_parameters(load_task_config, bq_pipeline_input_config):
+    path_exists = False
+    if 'path' in bq_pipeline_input_config:
+      # The path URL of the GCS load file(s).
+      load_task_config['path'] = bq_pipeline_input_config['path']
+      path_exists = True
+
+    table_exists = False
+    if 'table' in bq_pipeline_input_config:
+      # The destination bigquery table name for loading
+      load_task_config['table'] = bq_pipeline_input_config['table']
+      table_exists = True
+
+    schema_exists = False
+    if 'schema' in bq_pipeline_input_config:
+      # The schema of the destination bigquery table
+      load_task_config['schema'] = bq_pipeline_input_config['schema']
+    schema_exists = True
+
+    # We now figure out whether a load operation is required
+    if table_exists:
```

Comment:
Perhaps we can refactor it a bit to make it clearer? For example, 

if not table_exists and not  path_exists:
  return None
...

See if we can avoid nested ifs?

Reply:
In the process of doing this; stay tuned.

Reply:
So I simplified this somewhat, but I don't see a way in which it can be simplified further. I've added a unit-test for this that will hopefully make it clearer; PTAL.